plugins/legion-hid/fu-legion-hid-device.c

```diff
@@ -0,0 +1,658 @@
+	}
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 10, NULL);
+	for (guint i = 0; i < device_count; ++i)
+		fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 90 / device_count, NULL);
+
+	fu_progress_step_done(progress);
+
+	if (left_need_upgrade) {
+		if (!fu_legion_hid_device_execute_upgrade(self, img_left, error)) {
+			g_prefix_error_literal(error, "execute upgrade left gamepad failed: ");
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+		g_usleep(FU_LEGION_HID_DEVICE_REBOOT_WAIT_TIME * 1000);
+	}
+
+	if (right_need_upgrade) {
+		if (!fu_legion_hid_device_execute_upgrade(self, img_right, error)) {
+			g_prefix_error_literal(error, "execute upgrade right gamepad failed: ");
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+		g_usleep(FU_LEGION_HID_DEVICE_REBOOT_WAIT_TIME * 1000);
+	}
+
+	if (mcu_need_upgrade) {
+		if (!fu_legion_hid_device_execute_upgrade(self, img_mcu, error)) {
+			g_prefix_error_literal(error, "execute upgrade mcu failed: ");
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	} else {
+		fu_legion_hid_device_set_version(self, error);
+	}
+
+	return TRUE;
+}
+
+static void
+fu_legion_hid_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DECOMPRESSING, 0, "prepare-fw");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "reload");
+}
+
+static gboolean
+fu_legion_hid_device_validate_descriptor(FuDevice *device, GError **error)
+{
+	g_autoptr(FuHidDescriptor) descriptor = NULL;
+	g_autoptr(FuHidReport) report = NULL;
+
+	descriptor = fu_hidraw_device_parse_descriptor(FU_HIDRAW_DEVICE(device), error);
+	if (descriptor == NULL)
+		return FALSE;
+	report = fu_hid_descriptor_find_report(descriptor,
+					       error,
+					       "usage-page",
+					       0xFFA0,
+					       "usage",
+					       0x01,
+					       "collection",
+					       0x01,
+					       NULL);
+	if (report == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid_device_setup(FuDevice *device, GError **error)
+{
+	if (!fu_legion_hid_device_validate_descriptor(device, error))
+		return FALSE;
+
+	fu_legion_hid_device_set_version(FU_LEGION_HID_DEVICE(device), error);
+
+	return TRUE;
+}
+
+static FuFirmware *
+fu_legion_hid_device_prepare_firmware(FuDevice *device,
+				      GInputStream *stream,
+				      FuProgress *progress,
+				      FuFirmwareParseFlags flags,
+				      GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_legion_hid_firmware_new();
+
+	/* sanity check */
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+
+	return g_steal_pointer(&firmware);
+}
```

Comment:
if these are all the checks you're doing, you don't need this function at all if you use `fu_device_set_firmware_gtype()`